src/audio/buffer.c

```diff
@@ -80,21 +80,55 @@ struct comp_buffer *buffer_new(struct sof_ipc_buffer *desc)
 	assert(!memcpy_s(&buffer->ipc_buffer, sizeof(buffer->ipc_buffer),
 		       desc, sizeof(*desc)));
 
-	buffer->size = desc->size;
-	buffer->alloc_size = desc->size;
-	buffer->w_ptr = buffer->addr;
-	buffer->r_ptr = buffer->addr;
-	buffer->end_addr = buffer->addr + buffer->ipc_buffer.size;
-	buffer->free = buffer->ipc_buffer.size;
-	buffer->avail = 0;
-
-	buffer_zero(buffer);
+	buffer_init(buffer, desc->size);
 
 	spinlock_init(&buffer->lock);
 
 	return buffer;
 }
 
+int buffer_resize(struct comp_buffer *buffer, uint32_t size)
+{
+	struct sof_ipc_buffer *desc = NULL;
+
+	if (!buffer) {
+		trace_buffer_error("resize error: can't resize NULL buffer");
+		return -EINVAL;
+	}
+
+	/* validate request */
+	if (size == 0 || size > HEAP_BUFFER_SIZE) {
+		trace_buffer_error("resize error: size = %u is invalid", size);
+		return -EINVAL;
+	}
+
+	if (buffer->alloc_size == size)
+		return 0;
+
+	if (buffer->addr)
+		rfree(buffer->addr);
+
+	desc = &buffer->ipc_buffer;
+	if (!desc) {
+		trace_buffer_error("resize error: invalid buffer desc");
+		return -EINVAL;
+	}
+
+	buffer->addr = rballoc(RZONE_RUNTIME, desc->caps, size);
```

Comment:
Could we use `RZONE_BUFFER` to be consistent?
I know it doesn't matter, but still.